File: maths/sieve_of_eratosthenes.js
/**
 * Sieve of Eratosthenes
 */

function primeSieve(num) {
    if (num <= 0) return [];

    const sieve = Array(num + 1).fill(true);
    const primes = [];

    for (let p = 2; p * p <= num; p++) {
        if (sieve[p] === true) {
            for (let i = p * p; i <= num; i += p) {
                sieve[i] = false;
            }
        }
    }

    for (let p = 2; p <= num; p++) {
        if (sieve[p]) {
            primes.push(p);
        }
    }
    return primes;
}

const n = 50;
console.log(`Primes up to ${n}:`, primeSieve(n));

File: maths/sieve_of_eratosthenes.rs
pub fn prime_sieve(num: usize) -> Vec<usize> {
    if num == 0 {
        return vec![];
    }

    let mut sieve = vec![true; num + 1];
    let mut primes = Vec::new();

    let limit = (num as f64).sqrt() as usize;

    for p in 2..=limit {
        if sieve[p] {
            let mut i = p * p;
            while i <= num {
                sieve[i] = false;
                i += p;
            }
        }
    }

    for p in 2..=num {
        if sieve[p] {
            primes.push(p);
        }
    }
    primes
}

fn main() {
    let n = 50;
    println!("Primes up to {}: {:?}", n, prime_sieve(n));
}
